rust/blockstore/src/arrow/block/delta/builder_storage.rs

```diff
@@ -0,0 +1,161 @@
+use crate::{arrow::types::ArrowWriteableValue, key::CompositeKey};
+use either::Either;
+use std::collections::BTreeMap;
+
+pub struct BTreeBuilderStorage<V: ArrowWriteableValue> {
+    storage: BTreeMap<CompositeKey, V>,
+}
+
+impl<V: ArrowWriteableValue> BTreeBuilderStorage<V> {
+    fn add(&mut self, key: CompositeKey, value: V) {
+        self.storage.insert(key, value);
+    }
+
+    fn delete(&mut self, key: &CompositeKey) -> Option<V> {
+        self.storage.remove(key)
+    }
+
+    fn min_key(&self) -> Option<&CompositeKey> {
+        self.storage.keys().next()
+    }
+
+    fn split_off(&mut self, key: &CompositeKey) -> Self {
+        let split_off = self.storage.split_off(key);
+        Self { storage: split_off }
+    }
+
+    fn len(&self) -> usize {
+        self.storage.len()
+    }
+
+    fn iter<'referred_data>(
+        &'referred_data self,
+    ) -> Box<dyn Iterator<Item = (&'referred_data CompositeKey, &'referred_data V)> + 'referred_data>
+    {
+        Box::new(self.storage.iter())
+    }
+
+    fn into_iter(self) -> impl Iterator<Item = (CompositeKey, V)> {
+        self.storage.into_iter()
+    }
+}
+
+impl<V: ArrowWriteableValue> Default for BTreeBuilderStorage<V> {
+    fn default() -> Self {
+        Self {
+            storage: BTreeMap::new(),
+        }
+    }
+}
+
+/// This storage assumes that KV pairs are added in order. Deletes are a no-op. Calling `.add()` with the same key more than once is not allowed.
+pub struct VecBuilderStorage<V: ArrowWriteableValue> {
+    storage: Vec<(CompositeKey, V)>,
+}
+
+impl<V: ArrowWriteableValue> VecBuilderStorage<V> {
+    fn add(&mut self, key: CompositeKey, value: V) {
+        self.storage.push((key, value));
+    }
+
+    fn delete(&mut self, _: &CompositeKey) -> Option<V> {
+        None
+    }
+
+    fn min_key(&self) -> Option<&CompositeKey> {
+        self.storage.first().map(|(key, _)| key)
+    }
+
+    fn split_off(&mut self, key: &CompositeKey) -> Self {
+        let split_index = self.storage.binary_search_by(|(k, _)| k.cmp(key)).unwrap();
+        let split_off = self.storage.split_off(split_index);
+        self.storage.shrink_to_fit();
+        Self { storage: split_off }
+    }
+
+    fn len(&self) -> usize {
+        self.storage.len()
+    }
+
+    fn iter<'referred_data>(
+        &'referred_data self,
+    ) -> Box<dyn Iterator<Item = (&'referred_data CompositeKey, &'referred_data V)> + 'referred_data>
+    {
+        Box::new(self.storage.iter().map(|(k, v)| (k, v)))
```

Comment:
I don't understand what this map does since it seems like it returns the same thing as input?

Reply:
<img width="1068" alt="Screenshot 2024-11-05 at 1 52 43 PM" src="https://github.com/user-attachments/assets/387c5f2c-d805-4d55-a66d-a4079950b821">

I'll add a comment